packages/shared/src/api/lanyardApi.ts

```diff
@@ -60,15 +60,35 @@ export async function checkAttestedSignature(signData: string) {
   return false;
 }
 
-export async function discoverContacts(phoneNums: string[]) {
+export async function discoverContacts(
+  phoneNums: string[]
+): Promise<[string, string][]> {
   try {
-    const nums = diffContactBook(phoneNums);
-    // const lastSalt = formatUw('0');
-    const payload = ['whose-bulk', 0, nums.last, nums.add, nums.del];
-    const noun = dwim(payload);
-
+    const nums = await diffContactBook(phoneNums);
+    if (nums.add.toString() === '0' && nums.del.toString() === '0') {
+      logger.log('discoverContacts: no changes, returning empty');
+      logger.trackEvent(AnalyticsEvent.DebugContactMatching, {
+        context: 'discoverContacts',
+        diffSetLength: 0,
+        delSetLength: 0,
+        message: 'No changes, no need to send request',
+      });
+      return [];
+    }
+    const lastSalt = (await db.lastLanyardSalt.getValue()) ?? formatUw('0');
```

Comment:
The `lastSalt` in the default case is sending the string `'0w0'` instead of a number value. Not that it matters in the default case, but you should be fine just sending `0` (or `0n`).

The json string that you get and inject into the database directly (below, L128-129) is going to be a `@ux`-encoded number. You'll want to pass _the number value_, not its string representation, into `dwim`/`nounify`. Probably:

```suggestion
    const lastSalt = BigInt('0x' + parseUx((await db.lastLanyardSalt.getValue()) ?? '0x0'), 16);
```

(Would be nice if json supported `bigint`s, then we wouldn't be forced to do song-and-dance here... ): )